synapseclient/core/otel_config.py

```diff
@@ -0,0 +1,233 @@
+        """Propagates attributes from the child span back to the parent span"""
+        parent_span = get_current_span()
+        if parent_span is not None and parent_span.is_recording():
+            for attribute in self.attributes_to_propagate_to_parent:
+                child_val = span.attributes.get(attribute)
+                parent_val = parent_span.attributes.get(attribute)
+                if child_val and not parent_val:
+                    # Propagate the attribute back to parent span
+                    parent_span.set_attribute(attribute, child_val)
+
+    def shutdown(self) -> None:
+        """No-op method that does nothing when the span processor is shut down."""
+
+    def force_flush(self, timeout_millis: int = 30000) -> None:
+        """No-op method that does nothing when the span processor is forced to flush."""
+
+
+def configure_traces(
+    resource_attributes: Optional[Dict[str, Any]] = None,
+    include_context: bool = True,
+) -> TracerProvider:
+    """
+    Configure OpenTelemetry tracing for the Synapse Python Client.
+
+    Args:
+        resource_attributes: Additional resource attributes to include
+        include_context: Whether to include contextual information about the runtime environment
+
+    Returns:
+        The configured TracerProvider
+    """
+    resource_attrs = {
+        SERVICE_NAME: os.environ.get("OTEL_SERVICE_NAME", DEFAULT_SERVICE_NAME),
+        SERVICE_VERSION: CLIENT_VERSION,
+        SERVICE_INSTANCE_ID: os.environ.get(
+            "OTEL_SERVICE_INSTANCE_ID", "default_instance"
+        ),
+    }
+
+    if include_context:
+        resource_attrs["python.version"] = ".".join(
+            str(v) for v in sys.version_info[:3]
+        )
+
+        resource_attrs["os.type"] = os.name
+
+        try:
+            from synapseclient import __version__ as client_version
+
+            resource_attrs[SERVICE_VERSION] = client_version
+        except ImportError:
+            pass
+
+    if resource_attributes:
+        resource_attrs.update(resource_attributes)
+
+    resource = Resource.create(resource_attrs)
+
+    provider = TracerProvider(resource=resource)
+
+    attribute_propagator = AttributePropagatingSpanProcessor(
+        attributes_to_propagate_to_child=[
+            "synapse.transfer.direction",
+            "synapse.operation.category",
+        ]
+    )
+    provider.add_span_processor(attribute_propagator)
+
+    otlp_exporter = OTLPSpanExporter()
+    provider.add_span_processor(BatchSpanProcessor(otlp_exporter))
+
+    if os.environ.get("OTEL_DEBUG_CONSOLE", "").lower() in (
+        "true",
+        "1",
+    ):
+        provider.add_span_processor(BatchSpanProcessor(ConsoleSpanExporter()))
+
+    # Set as global tracer provider
+    trace.set_tracer_provider(provider)
+
+    return provider
+
+
+def configure_metrics(
+    resource_attributes: Optional[Dict[str, Any]] = None,
+    include_context: bool = True,
+) -> MeterProvider:
```

Comment:
I am not currently collecting any metrics in this code, however, I am leaving this in so we have a good starting point if/when we want to collect metrics.

Reply:
Can you provide an example of what a metric we could collect is in the future?

Reply:
Cpu/memory usage. Connection pooling for http connections.